zipline/pipeline/hooks/progress.py

```diff
@@ -0,0 +1,494 @@
+        return cls(publisher_factory=lambda: publisher)
+
+    # The state manages our ProgressModel because we receive the information
+    # necessary to initialize over the course of several hook callbacks.
+    @property
+    def _model(self):
+        return self._state.model
+
+    def _publish(self):
+        self._state.publish()
+
+    @contextmanager
+    def running_pipeline(self, pipeline, start_date, end_date):
+        self._state.set_bounds(start_date, end_date)
+        try:
+            yield
+        except Exception:
+            self._model.finish(success=False)
+            self._publish()
+            raise
+        else:
+            self._model.finish(success=True)
+            self._publish()
+        finally:
+            self._state.reset()
+
+    def on_create_execution_plan(self, plan):
+        self._state.set_execution_plan(plan)
+        self._publish()
+
+    @contextmanager
+    def computing_chunk(self, plan, start_date, end_date):
+        try:
+            self._model.start_chunk(start_date, end_date)
+            self._publish()
+            yield
+        finally:
+            self._model.finish_chunk(start_date, end_date)
+            self._publish()
+
+    @contextmanager
+    def loading_terms(self, terms):
+        try:
+            self._model.start_load_terms(terms)
+            self._publish()
+            yield
+        finally:
+            self._model.finish_load_terms(terms)
+            self._publish()
+
+    @contextmanager
+    def computing_term(self, term):
+        try:
+            self._model.start_compute_term(term)
+            self._publish()
+            yield
+        finally:
+            self._model.finish_compute_term(term)
+            self._publish()
+
+
+class _ProgressHooksState(object):
+    """
+    Helper class for managing incremental acquisition of pipeline state in
+    ``ProgressHooks``.
+    """
+
+    def __init__(self, publisher_factory):
+        self._publisher_factory = publisher_factory
+        self.reset()
+
+    def reset(self):
+        self._start_date = None
+        self._end_date = None
+        self._plan = None
+
+        self._model = None
+        self._publisher = None
+
+    def set_bounds(self, start_date, end_date):
+        self._start_date = start_date
+        self._end_date = end_date
+
+    def set_execution_plan(self, plan):
+        # NOTE: This will be called multiple times in a chunked execution, but
+        # the plans should always be functionally equivalent.
+        self._plan = plan
+
+    def publish(self):
+        self.publisher.publish(self.model)
+
+    @property
+    def model(self):
+        _model = self._model
+        if _model is None:
+            _model = self._model = self._create_model()
+        return _model
+
+    @property
+    def publisher(self):
+        publisher = self._publisher
+        if publisher is None:
+            publisher = self._publisher = self._publisher_factory()
+        return self._publisher
```

Comment:
I realize they are both the same value at this point. but did you mean for this to be `return publisher` because of the assignment above?

Reply:
yup

Reply:
Did this.